cpp/include/cudf/lists/detail/gather.cuh

```diff
@@ -0,0 +1,219 @@
+#pragma once
+
+#include <thrust/transform_scan.h>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/lists/lists_column_view.hpp>
+
+namespace cudf {
+namespace lists {
+namespace detail {
+
+/**
+ * @brief The information needed to create an iterator to gather level N+1
+ *
+ * See documentation for make_gather_offsets for a detailed explanation.
+ */
+struct gather_data {
+  size_type const* offsets;
+  std::unique_ptr<column> base_offsets;
+};
+
+/**
+ * @brief Generates the data needed to create a `gather_map` for the next level of
+ * recursion in a hierarchy of list columns.
+ *
+ * Gathering from a single level of a list column is similar to gathering from
+ * a string column.  Each row represents a list bounded by offsets. Example:
+ *
+ * Level 0 : List<List<int>>
+ *           Size : 3
+ *           Offsets : [0, 2, 5, 10]
+ *
+ * This represents a column with 3 rows.
+ * Row 0 has 2 elements (bounded by offsets 0,2)
+ * Row 1 has 3 elements (bounded by offsets 2,5)
+ * Row 2 has 5 eleemnts (bounded by offsets 5,10)
+ *
+ * If we wanted to gather rows 0 and 2 the offsets for our outgoing column
+ * would be the compacted ranges (0,2) and (5,10). The level 1 column
+ * then looks like
+ *
+ * Level 1 : List<int>
+ *           Size : 2
+ *           Offsets : [0, 2, 7]
+ *
+ * However, we need to then gather one level further, because at the bottom we have
+ * a column of integers.  We cannot gather the elements in the ranges (0, 2) and (2, 7).
+ * Instead, we have to gather elements in the ranges from the Level 0 column (0, 2) and (5, 10).
+ * So our gather_map iterator will need to know these "base" offsets to index properly.
+ * Specifically:
+ *
+ * Offsets        : [0, 2, 7]    The offsets for Level 1
+ * Base Offsets   : [0, 5]       The corresponding base offsets from Level 0
+ *
+ * Using this we can create an iterator that generates the sequence which properly indexes the
+ * final integer values we want to gather.
+ *
+ * [0, 1, 5, 6, 7, 8, 9]
+ *
+ * Thinking generally, this means that to produce a gather_map for level N+1, we need to use the
+ * offsets from level N and the "base" offsets from level N-1. So we are always carrying along
+ * one extra buffer of these "base" offsets which keeps our memory usage well controlled.
+ *
+ * A concrete example:
+ *
+ * "column of lists of lists of ints"
+ * {
+ *   {
+ *      {2, 3}, {4, 5}
+ *   },
+ *   {
+ *      {6, 7, 8}, {9, 10, 11}, {12, 13, 14}
+ *   },
+ *   {
+ *      {15, 16}, {17, 18}, {17, 18}, {17, 18}, {17, 18}
+ *   }
+ * }
+ *
+ * List<List<int32_t>>:
+ * Length : 3
+ * Offsets : 0, 2, 5, 10
+ * Children :
+ *    List<int32_t>:
+ *    Length : 10
+ *    Offsets : 0, 2, 4, 7, 10, 13, 15, 17, 19, 21, 23
+ *       Children :
+ *           2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 17, 18, 17, 18, 17, 18
```

Comment:
I don't know if you need to show both forms of the example input here. I certainly prefer the former but would want to show the results in the same format.